File: sources/Android/sdk/src/main/java/com/lua/mobile/sdk/engine/register/LuaAbstractClassRegister.java
/**
 * Android LuaMobileSdk for Android framework project.
 *
 * Copyright 2016 Arno Zhang <zyfgood12@163.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.lua.mobile.sdk.engine.register;

import android.support.annotation.NonNull;
import android.util.Log;
import com.lua.mobile.sdk.engine.LuaScriptEngine;
import com.lua.mobile.sdk.engine.args.LuaBridgeArgs;
import com.lua.mobile.sdk.engine.bridge.InvokeParamsExtractor;
import com.lua.mobile.sdk.engine.bridge.LuaAbstractBridge;
import com.lua.mobile.sdk.engine.bridge.LuaBridgedFunction;
import com.lua.mobile.sdk.engine.bridge.LuaClassBridge;
import com.lua.mobile.sdk.engine.bridge.LuaMethod;
import com.lua.mobile.sdk.engine.lua.LuaDeclares;
import com.lua.mobile.sdk.engine.lua.LuaState;
import com.lua.mobile.sdk.engine.script.LuaType;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;

public class LuaAbstractClassRegister extends AbstractRegister {

    private static final String TAG = "LuaClassRegister";


    protected LuaState mLuaState;


    public LuaAbstractClassRegister(LuaScriptEngine scriptEngine) {
        super(scriptEngine);
        mLuaState = scriptEngine.getLuaState();
    }

    /**
     * stack:  {ClassName}, {meta}
     */
    public <T> void registerClassPrepare(@NonNull final LuaClassBridge<T> bridge) {
        String name = bridge.getBridgeName();
        mScriptEngine.addBridgedClass(bridge.getDataClass(), name);

        // temp = {}
        mLuaState.newTable();

        // _G.ClassName = temp
        mLuaState.pushIndexValueToTop(-1);
        mLuaState.setGlobal(name);

        // register  vars.
        //
        // -> public final  var
        //
        Field[] fields = bridge.getClass().getFields();
        registerStaticVariableInternal(fields, bridge);

        // meta = {}
        mLuaState.newTable();

        // _G.ClassName.metatable = meta
        mLuaState.pushIndexValueToTop(-1);
        mLuaState.setMetaTable(-3);

        // meta.__java_object_type = JAVA_OBJECT
        mLuaState.pushNumber(LuaType.JAVA_OBJECT);
        mLuaState.setField(-2, LuaDeclares.__java_object_type);

        // register meta methods.
        // meta.__java_meta_funcs = {}
        mLuaState.newTable();
        bindMetaMethods(-2, bridge);
        mLuaState.setField(-2, LuaDeclares.__java_meta_funcs);

        afterBindMetaMethods(-1);
    }

    protected <T> void bindMetaMethods(int tableIndex, @NonNull final LuaClassBridge<T> bridge) {
        // meta
        final String name = bridge.getBridgeName();
        LuaBridgedFunction call = new LuaBridgedFunction(mScriptEngine) {

            @Override
            protected int invokeInternal(LuaBridgeArgs args) throws Exception {
                // instance = new()
                // instance.metatable = meta
                //
                T instance = bridge.createInstance(mScriptEngine, args);
                mLuaState.pushJavaObjectWithMeta(instance, name);

                return 1;
            }
        };

        // meta.__java_meta_funcs.__call = call
        mLuaState.pushBridgedFunction(call);
        call.setFunctionName(name + "." + LuaDeclares.__call);
        mLuaState.setField(tableIndex, LuaDeclares.__call);
    }

    protected void afterBindMetaMethods(int tableIndex) {
        // meta.__call = call
        mLuaState.bindMetaCall(tableIndex);
    }

    protected void registerStaticVariableInternal(Field[] fields, LuaAbstractBridge bridge) {
        registerStaticVariableInternal((LuaScriptEngine) mScriptEngine, fields, bridge);
    }

    /**
     * register  final variables.
     *
     * stack[-1]   -   Module / Class
     */
    public static void registerStaticVariableInternal(
            LuaScriptEngine scriptEngine, Field[] fields, LuaAbstractBridge bridge) {

        LuaState state = scriptEngine.getLuaState();
        for (Field field : fields) {
            if (field.isAnnotationPresent(LuaMethod.class)) {
                continue;
            }

            int modifiers = field.getModifiers();
            if (!Modifier.isPublic(modifiers)
                    || !Modifier.isFinal(modifiers)
                    || !Modifier.isStatic(modifiers)) {
                continue;
            }

            // _G.ModuleName.fieldName = field
            try {
                state.push(field.get(bridge));
                state.setField(-2, field.getName());
            } catch (IllegalAccessException e) {
                Log.e(TAG, "registerStaticVariableInternal", e);
            }
        }
    }

    protected void registerBridgedFunctionInternal(Method[] methods, LuaAbstractBridge bridge) {
        registerBridgedFunctionInternal((LuaScriptEngine) mScriptEngine, methods, bridge);
    }

    protected static void registerBridgedFunctionInternal(
            LuaScriptEngine scriptEngine, Method[] methods, LuaAbstractBridge bridge) {

        LuaState state = scriptEngine.getLuaState();

        for (final Method method : methods) {
            int modifiers = method.getModifiers();
            if (!Modifier.isPublic(modifiers)
                    || !method.isAnnotationPresent(LuaMethod.class)) {
                continue;
            }

            String name = method.getName();
            LuaBridgedFunction function = createBridgedFunction(scriptEngine, method, bridge);

            // meta.method = function
            function.setFunctionName(bridge.getBridgeName() + "." + name);
            state.pushBridgedFunction(function);
            state.setField(-2, name);
        }
    }

    private static LuaBridgedFunction createBridgedFunction(
            LuaScriptEngine scriptEngine,
            @NonNull final Method method, @NonNull final Object bridge) {

        Class<?> retClazz = method.getReturnType();
        final boolean hasReturnType = retClazz != null && retClazz != void.class && retClazz != Void.class;
        final InvokeParamsExtractor extractor = InvokeParamsExtractor.createParamsExtractor(method);

        final LuaState mLuaState = scriptEngine.getLuaState();
        return new LuaBridgedFunction(scriptEngine) {
            @Override
            protected LuaBridgeArgs createBridgeArgs() {
                return new LuaBridgeArgs(mScriptEngine);
            }

            @Override
            protected int invokeInternal(LuaBridgeArgs bridgeArgs) throws Throwable {
                Object returnValue = null;

                try {
                    if (extractor == null) {
                        // R function()
                        returnValue = method.invoke(bridge);
                    } else {
                        returnValue = method.invoke(bridge, extractor.extractParams(bridgeArgs));
                    }
                } catch (Throwable e) {
                    Log.e(TAG, String.format("Invoke bridge member-method FAILED! target = %s, method = %s",
                            bridge.getClass().getSimpleName(), method.getName()), e);
                }

                if (hasReturnType) {
                    mLuaState.push(returnValue);
                    return 1;
                }

                return 0;
            }
        };
    }
}

File: sources/iOS/LuaMobileSdk/LuaMobileSdk/Sources/Api/engine/register/LuaAbstractClassRegister.m
/**
 * Android LuaMobileSdk for iOS framework project.
 *
 * Copyright 2017 Arno Zhang <zyfgood12@163.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "LuaType.h"
#import "LuaDeclares.h"
#import "LuaBridgeUtils.h"
#import "LuaState.h"
#import "LuaBridgeArgs.h"
#import "LuaScriptEngine.h"
#import "LuaClassBridge.h"
#import "LuaAbstractClassRegister.h"


int sdk_MethodInvokerInternal(lua_State* ls, BOOL isStatic) {
    MethodInvokeWrapper* wrapper = sdk_getInvokerWrapper(ls);
    if (!wrapper) {
        return 0;
    }

    LuaState* state = (__bridge LuaState*) wrapper->luaState;
    LuaMethodInvokeEnv* env = (__bridge LuaMethodInvokeEnv*) wrapper->env;

    if (isStatic) {
        // 占位.
        [state pushNil];
        [state insert:1];
        [state pushNil];
        [state insert:1];
    }

    LuaScriptEngine* engine = [state getScriptEngine];
    LuaBridgeArgs* args = [[LuaBridgeArgs alloc] init:engine];
    [state push:(*wrapper->method)(env, args)];

    return 1;
}

int sdk_instanceMethodInvoker(lua_State* ls) {
    return sdk_MethodInvokerInternal(ls, NO);
}

int sdk_staticMethodInvoker(lua_State* ls) {
    return sdk_MethodInvokerInternal(ls, YES);
}

int sdk_callInvoker(lua_State* ls) {
    MethodInvokeWrapper* wrapper = sdk_getInvokerWrapper(ls);
    if (!wrapper) {
        return 0;
    }

    LuaState* state = (__bridge LuaState*) wrapper->luaState;
    LuaMethodInvokeEnv* env = (__bridge LuaMethodInvokeEnv*) wrapper->env;
    LuaClassBridge* bridge = (LuaClassBridge*) env.bridge;
    LuaScriptEngine* engine = [state getScriptEngine];

    LuaBridgeArgs* args = [[LuaBridgeArgs alloc] init:engine];
    NSObject* instance = [bridge createInstance:engine Args:args];
    [state pushObjectWithMeta:instance ClassName:[bridge getBridgeName]];

    return 1;
}


@implementation LuaAbstractClassRegister


- (instancetype)initWithEngine:(LuaScriptEngine*)engine
                  ClassManager:(id <RegisterClassManager>)manager {
    if (self = [super initWithEngine:engine]) {
        mLuaState = [engine getLuaState];
        mRegisterClassManager = manager;
    }

    return self;
}

- (BOOL)registerClassPrepare:(LuaClassBridge*)bridge {
    bridge.env = [[LuaMethodInvokeEnv alloc] init:bridge Reg:self];
    [self holdBridge:bridge];

    NSString* name = [bridge getBridgeName];
    if (!name || name.length <= 0) {
        @throw [[NSException alloc]
                initWithName:@"GlobalClassRegister"
                      reason:@"LuaClassBridge name must not empty!" userInfo:nil];
        return NO;
    }

    [mRegisterClassManager addBridgedClass:[bridge getDataClass] ClassName:name];

    // temp = {}
    [mLuaState newTable];

    // _G.ClassName = temp
    [mLuaState pushIndexValueToTop:-1];
    [mLuaState setGlobal:name];

    // register const static vars.
    [bridge registerStaticVariables:self];

    // register static methods.
    [bridge registerStaticMethods:self];

    // meta = {}
    [mLuaState newTable];

    // _G.ClassName.metatable = meta
    [mLuaState pushIndexValueToTop:-1];
    [mLuaState setMetaTable:-3];

    // meta.__oc_object_type = LuaTypeObject
    [mLuaState pushNumber:LuaTypeObject];
    [mLuaState setField:-2 CStrKey:__oc_object_type];

    // register meta methods.
    [self bindMetaMethods:bridge];

    return YES;
}

- (void)bindMetaMethods:(LuaClassBridge*)bridge {
    // meta.__call = call
    [self newWrappedClosureAtTop:bridge Method:nil Closure:&sdk_callInvoker];
    [mLuaState setField:-2 CStrKey:__call];
}

- (void)registerBool:(NSString*)name Value:(BOOL)value {
    [mLuaState pushBool:value];
    [mLuaState setField:-2 Key:name];
}

- (void)registerInt:(NSString*)name Value:(int)value {
    [mLuaState pushInt:value];
    [mLuaState setField:-2 Key:name];
}

- (void)registerLong:(NSString*)name Value:(long)value {
    [self registerDouble:name Value:value];
}

- (void)registerFloat:(NSString*)name Value:(float)value {
    [self registerDouble:name Value:value];
}

- (void)registerDouble:(NSString*)name Value:(double)value {
    [mLuaState pushNumber:value];
    [mLuaState setField:-2 Key:name];
}

- (void)registerString:(NSString*)name Value:(NSString*)value {
    [mLuaState pushString:value];
    [mLuaState setField:-2 Key:name];
}

- (void)registerObject:(NSString*)name Value:(NSObject*)value {
    [mLuaState push:value];
    [mLuaState setField:-2 Key:name];
}

- (void)registerBridgedFunction:(NSString*)name Value:(LuaBridgedFunction*)value {
    [mLuaState pushBridgedFunction:value];
    [mLuaState setField:-2 Key:name];
}

- (void)registerInstanceMethod:(LuaAbstractBridge*)bridge
                          Name:(NSString*)name
                        Method:(StaticBridgeMethod)method {

    [self newWrappedClosureAtTop:bridge Method:method Closure:&sdk_instanceMethodInvoker];
    [mLuaState setField:-2 Key:name];
}

- (void)registerStaticMethod:(LuaAbstractBridge*)bridge
                        Name:(NSString*)name
                      Method:(StaticBridgeMethod)method {

    [self newWrappedClosureAtTop:bridge Method:method Closure:&sdk_staticMethodInvoker];
    [mLuaState setField:-2 Key:name];
}

- (void)newWrappedClosureAtTop:(LuaAbstractBridge*)bridge
                        Method:(void*)method
                       Closure:(lua_CFunction)closure {

    LuaWithMemberBridge* module = (LuaWithMemberBridge*) bridge;

    MethodInvokeWrapper* wrapper = [mLuaState newUserData:sizeof(MethodInvokeWrapper)];
    wrapper->method = method;
    wrapper->env = (__bridge void*) module.env;
    wrapper->luaState = (__bridge void*) mLuaState;

    [mLuaState pushCClosure:closure UpValueCount:1];
}

@end